tools/testing/e2e/iree-e2e-attention-test.cc

```diff
@@ -0,0 +1,476 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "iree/base/api.h"
+#include "iree/base/internal/cpu.h"
+#include "iree/base/internal/flags.h"
+#include "iree/base/internal/math.h"
+#include "iree/base/internal/path.h"
+#include "iree/hal/api.h"
+#include "iree/modules/hal/module.h"
+#include "iree/tooling/context_util.h"
+#include "iree/tooling/device_util.h"
+#include "iree/vm/api.h"
+#include "iree/vm/native_module_cc.h"
+#include "tools/testing/e2e/test_utils.h"
+
+//===----------------------------------------------------------------------===//
+// Reference Attention
+//===----------------------------------------------------------------------===//
+
+// Helper for reference_attention.
+// Function to allocate and initialize tensors
+float* allocate_tensor(int dim1, int dim2, int dim3) {
+  float* tensor = (float*)malloc(dim1 * dim2 * dim3 * sizeof(float));
+  for (int i = 0; i < dim1 * dim2 * dim3; ++i) {
+    tensor[i] = 0.0f;
+  }
+  return tensor;
+}
+
+// Function to free allocated tensors
+void free_tensor(float* tensor) { free(tensor); }
+
+// Function to calculate 1D index for a 3D array
+int index_3d(int i, int j, int k, int dim2, int dim3) {
+  return i * dim2 * dim3 + j * dim3 + k;
+}
+
+static void reference_attention_f32_f32_f32_f32(
+    iree_hal_dim_t M, iree_hal_dim_t K1, iree_hal_dim_t K2, iree_hal_dim_t N,
+    iree_hal_dim_t B, const float* query_data, const float* key_data,
+    const float* value_data, float* result_data, iree_hal_dim_t b) {
+  float* Attention = allocate_tensor(B, M, K2);
+
+  // Compute Q * K^T
+  for (int m = 0; m < M; ++m) {
+    for (int k2 = 0; k2 < K2; ++k2) {
+      float sum = 0.0;
+      for (int k1 = 0; k1 < K1; ++k1) {
+        int q_idx = index_3d(b, m, k1, M, K1);
+        int k_idx = index_3d(b, k2, k1, K2, K1);
+        sum += query_data[q_idx] * key_data[k_idx];
+      }
+      int att_idx = index_3d(b, m, k2, M, K2);
+      Attention[att_idx] = sum / sqrt(K1);  // Scale by sqrt(K1)
+    }
+  }
```

Comment:
ASan is flagging some issues with this reference code: https://github.com/iree-org/iree/actions/runs/10355617561/job/28663687393?pr=17751#step:4:12011
```
  1064/1908 Test  #996: iree/tests/e2e/attention/e2e_attention_cpu_f16_f16_f16_small_llvm-cpu_local-task ................................***Failed    1.86 sec
  --- TEST[attention_1_128_64_512_32_dtype_f16_f16_f16_f16_1_128_64_512_32_64_1.0_0] ---
  Attention shape (BATCHxMxK1xK2xN): 1x128x64x512x64x32
  =================================================================
  ==67947==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x531000038800 at pc 0x561da0238b21 bp 0x7ffd4efe6350 sp 0x7ffd4efe6348
  READ of size 4 at 0x531000038800 thread T0
      #0 0x561da0238b20 in reference_attention_f32_f32_f32_f32(unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, float const*, float const*, float const*, float*, unsigned long) /work/tools/testing/e2e/iree-e2e-attention-test.cc:60:36
      #1 0x561da0238b20 in reference_attention_element(unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned int, unsigned int, unsigned int, void*, void*, void*, void*, void*, unsigned long) /work/tools/testing/e2e/iree-e2e-attention-test.cc:114:5
      #2 0x561da0238b20 in reference_attention(unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned int, unsigned int, unsigned int, iree_byte_span_t, iree_byte_span_t, iree_byte_span_t, iree_byte_span_t, iree_byte_span_t, int) /work/tools/testing/e2e/iree-e2e-attention-test.cc:147:5
      #3 0x561da0238b20 in check_attention_results_impl(_IO_FILE*, attention_results_t const*, int) /work/tools/testing/e2e/iree-e2e-attention-test.cc:293:3
      #4 0x561da0238b20 in check_attention_results(_IO_FILE*, attention_results_t const*) /work/tools/testing/e2e/iree-e2e-attention-test.cc:315:7
      #5 0x561da0238b20 in iree::AttentionTestModuleState::CheckAttentionResults(iree::vm::ref<iree_hal_device_t>, long, long, long, long, long, iree::vm::ref<iree_hal_buffer_view_t>, iree::vm::ref<iree_hal_buffer_view_t>, iree::vm::ref<iree_hal_buffer_view_t>, iree::vm::ref<iree_hal_buffer_view_t>) /work/tools/testing/e2e/iree-e2e-attention-test.cc:410:28
```

Reply:
We suspect this is due to the lack of the proper type conversion for the reference implementation. I will see the result, and then I will double-check it.

Reply:
Solved that problem. Proper conversion from F16 to F32 worked out.